wcfsetup/install/files/js/WoltLabSuite/Core/Ui/File/Upload.js

```diff
@@ -0,0 +1,245 @@
+	"use strict";
+	
+	/**
+	 * @constructor
+	 */
+	function Upload(buttonContainerId, targetId, options) {
+		options = options || {};
+		
+		if (options.internalId === undefined) {
+			throw new Error("Missing internal id.");
+		}
+		
+		// set default options
+		this._options = Core.extend({
+			// name if the upload field
+			name: '__files[]',
+			// is true if every file from a multi-file selection is uploaded in its own request
+			singleFileRequests: false,
+			// url for uploading file
+			url: 'index.php?ajax-file-upload/&t=' + SECURITY_TOKEN,
+			// image preview
+			imagePreview: false,
+			// max files
+			maxFiles: null
+		}, options);
+		
+		this._options.multiple = this._options.maxFiles === null || this._options.maxFiles > 1; 
+		
+		this._options.url = Core.convertLegacyUrl(this._options.url);
+		if (this._options.url.indexOf('index.php') === 0) {
+			this._options.url = WSC_API_URL + this._options.url;
+		}
+		
+		this._buttonContainer = elById(buttonContainerId);
+		if (this._buttonContainer === null) {
+			throw new Error("Element id '" + buttonContainerId + "' is unknown.");
+		}
+		
+		this._target = elById(targetId);
+		if (targetId === null) {
+			throw new Error("Element id '" + targetId + "' is unknown.");
+		}
+		
+		if (options.multiple && this._target.nodeName !== 'UL' && this._target.nodeName !== 'OL') {
+			throw new Error("Target element has to be list or table body if uploading multiple files is supported.");
+		}
+		
+		this._fileElements = [];
+		this._internalFileId = 0;
+		
+		// upload ids that belong to an upload of multiple files at once
+		this._multiFileUploadIds = [];
+		
+		this._createButton();
+		this.checkMaxFiles();
+		
+		this._deleteHandler = new DeleteHandler(buttonContainerId, targetId, this._options.imagePreview, this);
+	}
+	
+	Core.inherit(Upload, CoreUpload, {
+		_createFileElement: function(file) {
+			var element = Upload._super.prototype._createFileElement.call(this, file);
+			element.classList.add('box64', 'uploadedFile');
+			
+			var progress = elBySel('progress', element);
+			
+			var icon = elCreate('span');
+			icon.classList = 'icon icon64 fa-spinner';
+			
+			var fileName = element.textContent;
+			element.textContent = "";
+			element.append(icon);
+			
+			var innerDiv = elCreate('div');
+			var fileNameP = elCreate('p');
+			fileNameP.textContent = fileName; // file.name
+			
+			var smallProgress = elCreate('small');
+			smallProgress.appendChild(progress);
+			
+			innerDiv.appendChild(fileNameP);
+			innerDiv.appendChild(smallProgress);
+			
+			var div = elCreate('div');
+			div.appendChild(innerDiv);
+			
+			var ul = elCreate('ul');
+			ul.classList = 'buttonGroup';
+			div.appendChild(ul);
+			
+			// reset element textContent and replace with own element style
+			element.append(div);
+			
+			return element;
+		},
+		
+		_failure: function(uploadId, data, responseText, xhr, requestOptions) {
+			for (var i in this._fileElements[uploadId]) {
```

Comment:
Is `this._fileElements` a continuous array? Use a normal `for`-loop then instead of `for (… in …)`.